### ISSUE TITLE
xtb crashes on multiple architectures

### ISSUE THREAD

**Issue by susilehtola** [grimme-lab/xtb#267]
There are crashes in the test suite on multiple architectures. The tests run correctly on

- x86_64
- aarch64
- ppc64le

but they fail on
- i686
- armv7hl
- s390x

Here's are the i686 build logs for reference.

[root.log](https://github.com/grimme-lab/xtb/files/4797942/root.log)
[build.log](https://github.com/grimme-lab/xtb/files/4797943/build.log)

**Comment by awvwgk** [comment]
Thanks for testing, it's nice to know we can run on aarch64 and ppc64le, since we usually only test on x86_64 architectures.

Looks like the problem is a the call to an `external` declared subroutine, so this _might_ go away by adding an explicit interface for `generate_wsc` (or packing it in a module). If it is an implicit interface problem, it will take a significant effort to wrap everything in modules, still this might be worth the effort anyway.

Also, I'm missing the resources and time to test for other architectures than x86_64, if we can come up with a CI solution I'm will try to keep them running once they work, but it's rather unlikely that we (which mean me) will invest significant efforts to get 32 bit or arm architectures running with `xtb`.

**Comment by susilehtola** [comment]
It looks as if there is some memory error.

**Comment by susilehtola** [comment]
fml now there's a circular dependency error. Someone who actually understands the code should have a look at how it should be interfaced properly.

**Comment by awvwgk** [comment]
I guess that was the reason to make it `external` in the first place.

**Comment by susilehtola** [comment]
Well, I already tried to do that:
```
diff --git a/src/type/molecule.f90 b/src/type/molecule.f90
index fe0c6b2..02e2724 100644
--- a/src/type/molecule.f90
+++ b/src/type/molecule.f90
@@ -195,6 +195,14 @@ contains
 !> Constructor for the molecular structure type
 subroutine initMolecule &
      & (mol, at, sym, xyz, chrg, uhf, lattice, pbc)
+
+   interface
+      subroutine generate_wsc(mol,wsc)
+        type(TMolecule), intent(inout) :: mol
+        type(tb_wsc),    intent(inout) :: wsc
+      end subroutine generate_wsc
+   end interface
+
    type(TMolecule), intent(out) :: mol
    integer, intent(in) :: at(:)
    character(len=*), intent(in) :: sym(:)
```
and here's what I get:
```
[  0%] Building Fortran object CMakeFiles/xtb-object.dir/src/type/molecule.f90.o
f951: Warning: ‘-Werror=’ argument ‘-Werror=format-security’ is not valid for Fortran
/home/work/xtb.susi/src/type/molecule.f90:201:24:

  201 |         type(TMolecule), intent(inout) :: mol
      |                        1
Error: Derived type ‘tmolecule’ at (1) is being used before it is defined
/home/work/xtb.susi/src/type/molecule.f90:202:21:

  202 |         type(tb_wsc),    intent(inout) :: wsc
      |                     1
Error: Derived type ‘tb_wsc’ at (1) is being used before it is defined
/home/work/xtb.susi/src/type/molecule.f90:270:34:

  270 |    call generate_wsc(mol, mol%wsc)
      |                                  1
Error: Type mismatch in argument ‘mol’ at (1); passed TYPE(tmolecule) to INTEGER(4)
```

**Comment by awvwgk** [comment]
You forgot to import the datatypes in the interface (it's a different scope and therefore independent from the function body):
```fortran
   interface
      subroutine generate_wsc(mol,wsc)
        import :: TMolecule, tb_wsc
        type(TMolecule), intent(inout) :: mol
        type(tb_wsc),    intent(inout) :: wsc
      end subroutine generate_wsc
   end interface

**Comment by susilehtola** [comment]
Ah, cool, thanks

**Comment by susilehtola** [comment]
#269 adds the interfaces, but I can confirm that the tests still fail on i686 by rebuilding the package in a 32-bit chroot environment.

**Comment by awvwgk** [comment]
Now the more important information would be if the previous crash is resolved with this.

**Comment by susilehtola** [comment]
@awvwgk I found the problem and added the fix in #269. The problem is that `c=0` in some loops, which causes a problem in the array splicing.

**Comment by susilehtola** [comment]
With #269, the tests succeed also on i686, armv7hl, s390x.
